### PR TITLE
Chassis update

### PR DESCRIPTION
**Pull request by outsidermm**
Rewrite chassis.py and physics.py using Pheonix 6 API and maintain TalonSRX support through Pheonix 5

### CODE REVIEW

File: physics.py [hunk 1]
```diff
@@ -16,22 +16,24 @@
...
-        self.sim_collection.setIntegratedSensorVelocity(int(velocity_cps / 10))
-        self.sim_collection.addIntegratedSensorPosition(int(velocity_cps * dt))
+        self.sim_collection.set_rotor_velocity(int(velocity_cps))
+        self.sim_collection.add_rotor_position(int(velocity_cps * dt))
 
 
 class SimpleTalonSRXMotorSim:
     def __init__(
-        self, motor: phoenix5.TalonSRX, kV: float, rev_per_unit: float
+        self, motor: phoenix6.hardware.TalonSRX, kV: float, rev_per_unit: float
```
Comment (auscompgeek) [inline physics.py:36]: I thought the Talon SRX isn't a Phoenix 6 device?

File: components/chassis.py [hunk 2]
```diff
@@ -61,67 +58,75 @@ def __init__(
...
-        self.drive.config_kD(0, 0, 10)
+        drive_config = self.drive.configurator
+
+        drive_motor_config = phoenix6.configs.MotorOutputConfigs()
+        drive_motor_config.neutral_mode = phoenix6.signals.NeutralModeValue.BRAKE
+        drive_motor_config.inverted = steer_reversed
+
+        self.drive_pid_ff = (
+            phoenix6.configs.Slot1Configs()
+        )  # configuration for motor pid and feedforward
```
Comment (auscompgeek) [inline components/chassis.py:94]: Comment on the line before, not this weird formatting please.

File: components/chassis.py [hunk 3]
```diff
@@ -61,67 +58,75 @@ def __init__(
...
     def get_angle_absolute(self) -> float:
         """Gets steer angle (radians) from absolute encoder"""
-        return math.radians(self.encoder.getAbsolutePosition())
+        return math.radians(self.encoder.get_absolute_position())
 
     def get_angle_integrated(self) -> float:
         """Gets steer angle from motor's integrated relative encoder"""
-        return self.steer.getSelectedSensorPosition() * self.STEER_COUNTS_TO_RAD
+        return self.steer.get_position() * self.STEER_COUNTS_TO_RAD
+        # return self.steer.getSelectedSensorPosition() * self.STEER_COUNTS_TO_RAD
```
Comment (auscompgeek) [inline components/chassis.py:116]:
```suggestion
```

File: components/chassis.py [hunk 4]
```diff
@@ -61,67 +58,75 @@ def __init__(
...
+
+        self.drive_pid_ff = (
+            phoenix6.configs.Slot1Configs()
+        )  # configuration for motor pid and feedforward
+        self.drive_pid_ff.k_p = 0.011489
+        self.drive_pid_ff.k_i = 0
+        self.drive_pid_ff.k_d = 0
+        self.drive_pid_ff.k_s = 0.18877
+        self.drive_pid_ff.k_v = 2.7713
+        self.drive_pid_ff.k_a = 0.18824
```
Comment (auscompgeek) [inline components/chassis.py:100]: Doesn't the config constructor take these as keyword arguments?

File: components/chassis.py [hunk 5]
```diff
@@ -61,67 +58,75 @@ def __init__(
...
-        self.drive.config_kI(0, 0, 10)
-        self.drive.config_kD(0, 0, 10)
+        drive_config = self.drive.configurator
+
+        drive_motor_config = phoenix6.configs.MotorOutputConfigs()
+        drive_motor_config.neutral_mode = phoenix6.signals.NeutralModeValue.BRAKE
+        drive_motor_config.inverted = steer_reversed
+
+        self.drive_pid_ff = (
+            phoenix6.configs.Slot1Configs()
```
Comment (auscompgeek) [inline components/chassis.py:93]: Why slot 1 instead of slot 0? Don't the control requests default to slot 0?

File: physics.py (hunk 1 above)
Comment (outsidermm) [reply physics.py:36]: Yes, TalonSRX is pheonix5 only. I just changed all occurrences of pheonix5 to pheonix6. Haven't got time to change the SRX yet.

File: components/chassis.py (hunk 4 above)
Comment (auscompgeek) [inline components/chassis.py:100]: The feedforward constants used metres as the distance base unit, but the PID gains used Talon units. Are these in the correct units for what Phoenix 6 expects?

File: components/chassis.py [hunk 6]
```diff
@@ -61,67 +58,75 @@ def __init__(
...
 
         # Reduce CAN status frame rates before configuring
-        self.steer.setStatusFramePeriod(
-            phoenix5.StatusFrameEnhanced.Status_1_General, 250, 10
-        )
-        self.drive.setStatusFramePeriod(
-            phoenix5.StatusFrameEnhanced.Status_1_General, 250, 10
-        )
+        self.steer.get_fault_field().set_update_frequency(4, 0.01)
+        self.drive.get_fault_field().set_update_frequency(4, 0.01)
```
Comment (auscompgeek) [inline components/chassis.py:68]: 1. Are we updating the correct frame(s) we don't care about?
2. Is the frame rate correct?

File: components/chassis.py [hunk 7]
```diff
@@ -61,67 +58,75 @@ def __init__(
...
 
     def get_rotation(self) -> Rotation2d:
         """Get the steer angle as a Rotation2d"""
         return Rotation2d(self.get_angle_integrated())
 
     def get_speed(self) -> float:
         # velocity is in counts / 100ms, return in m/s
-        return self.drive.getSelectedSensorVelocity() * self.DRIVE_COUNTS_TO_METRES * 10
+        return self.drive.get_velocity() * self.DRIVE_COUNTS_TO_METRES * 10
+        # return self.drive.getSelectedSensorVelocity() * self.DRIVE_COUNTS_TO_METRES * 10
```
Comment (auscompgeek) [inline components/chassis.py:125]:
```suggestion
```

File: components/chassis.py [hunk 8]
```diff
@@ -61,67 +58,75 @@ def __init__(
...
 
     def get_distance_traveled(self) -> float:
-        return self.drive.getSelectedSensorPosition() * self.DRIVE_COUNTS_TO_METRES
+        return self.drive.get_position() * self.DRIVE_COUNTS_TO_METRES
+        # return self.drive.getSelectedSensorPosition() * self.DRIVE_COUNTS_TO_METRES
```
Comment (auscompgeek) [inline components/chassis.py:129]:
```suggestion
```

File: components/chassis.py [hunk 9]
```diff
@@ -135,33 +140,33 @@ def set(self, desired_state: SwerveModuleState):
         self.state = SwerveModuleState.optimize(self.state, self.get_rotation())
 
         if abs(self.state.speed) < 0.01 and not self.module_locked:
-            self.drive.set(phoenix5.ControlMode.Velocity, 0)
-            self.steer.set(phoenix5.ControlMode.PercentOutput, 0)
+            drive_reqest = phoenix6.controls.VelocityVoltage(0).with_slot(1)
+            self.drive.set_control(drive_reqest)
```
Comment (auscompgeek) [inline components/chassis.py:144]: speeling

File: components/chassis.py [hunk 10]
```diff
@@ -135,33 +140,33 @@ def set(self, desired_state: SwerveModuleState):
...
         target_displacement = constrain_angle(
             self.state.angle.radians() - current_angle
         )
         target_angle = target_displacement + current_angle
-        self.steer.set(
-            phoenix5.ControlMode.Position, target_angle * self.STEER_RAD_TO_COUNTS
+        steer_request = (
+            phoenix6.controls.PositionVoltage()
+            .with_velocity(target_angle * self.STEER_RAD_TO_COUNTS)
+            .with_slot(0)
```
Comment (auscompgeek) [inline components/chassis.py:158]: Isn't slot 0 the default?

If it isn't, shouldn't we be preallocating the control request like CTRE's examples in the docs, and only updating what changes here?

**Review by auscompgeek — CHANGES_REQUESTED** [review]

File: components/chassis.py (hunk 6 above)
Comment (outsidermm) [reply components/chassis.py:68]: I don't know what is the equivalent of ctre.StatusFrameEnhanced.Status_1_General in pheonix6 but I think the get_fault_field() basically gives general status. 

The frame rate is correct, 4Hz. The previous code was 250 ms per refrash so 4 Hz

File: components/chassis.py (hunk 4 above)
Comment (outsidermm) [reply components/chassis.py:100]: the p,i,d,s,v,a aren't inputted in the parameters of the constructor if thats what you mean.

File: components/chassis.py (hunk 10 above)
Comment (outsidermm) [reply components/chassis.py:158]: It is, thats mb. well this steer_request is redefined by what changes here, the original code changes the control mode and the velocity

File: components/chassis.py (hunk 4 above)
Comment (auscompgeek) [reply components/chassis.py]: Huh, really? That's dumb. Dunno why they wouldn't allow that.

File: components/chassis.py [hunk 11]
```diff
@@ -396,7 +417,7 @@ def get_tilt_rate(self) -> float:
 
     @feedback
     def get_drive_current(self) -> float:
-        return sum(abs(x.drive.getStatorCurrent()) for x in self.modules)
+        return sum(abs(x.drive.get_stator_current().value) for x in self.modules)
```
Comment (auscompgeek) [inline components/chassis.py:420]: Eep, this is a code smell. The `Chassis` component should never have been reaching into the `SwerveModule` internals in the first place.

nit: add a `SwerveModule.get_drive_current() -> float` method

**Comment by outsidermm** [comment]
Is the sim working for you guys? Amber says navx is having issues, but I'm having issues with the sim not showing the TalonFX objects on one of its tabs

File: components/chassis.py (hunk 4 above)
Comment (outsidermm) [reply components/chassis.py]: Don't we have to retune the pid and the ff anyways

Comment (auscompgeek) [reply components/chassis.py]: Yes, but we shouldn't destroy the modules because we had nonsense numbers either.

Comment (outsidermm) [reply components/chassis.py]: alr I have converted the PID value to the new unit and I have reverted back to using wpimath ff as with the talon config ff idk how to even apply the ff values

File: robot.py [hunk 12]
```diff
@@ -46,16 +46,53 @@ def teleopInit(self) -> None:
...
+            self.gamepad.getLeftY(),
+            self.gamepad.getLeftX(),
+            self.gamepad.getRightX(),
+        ]
+
+        pov = self.gamepad.getPOV()
+        # print(pov)
+        if pov != -1:
+            joysticks[0] = 0
+            joysticks[1] = 0
```
Comment (auscompgeek) [inline robot.py:60]: What's with this `joysticks` list?

**Review by auscompgeek — APPROVED** [review]
LGTM

**Pull request merged**